Create a model registry with DVC and DAGsHub (P2)

deanp70 opened the pull request:
This PR enables Mindmeld users to create a central model registry on DAGsHub, making the model easily accessible and shareable with other team members by hosting all models in a central hub.

This refers to issue #300, which I discussed with @vijay120 and @alhuang10.

I added the option:
`python -m home_assistant dvc --setup_dagshub`

This defines the DVC remote as DAGsHub and guides the user in creating a project and providing the required information (URL, username, and password/access token).

I also edited the documentation to cover this change, and as far as I could test everything is passing.
Looking forward to getting your input on this :)

deanp70 commented:
@vijay120 gentle reminder. Would appreciate your input on this 🙏

vijay120 commented on mindmeld/components/classifier.py:
```diff
@@ -19,7 +19,7 @@
 from abc import ABC, abstractmethod
 
-from sklearn.externals import joblib
+import joblib
```
We have to use the older version of joblib due to backwards compatability issues with older models. Is this a required change for Dagshub?

deanp70 replied:
Hey @vijay120, no, I don't think it does. My setup was complaining so I changed it (mistakenly thought it would be an improvement), but admittedly I wasn't thinking about backward compatibility issues. Will undo it and commit.

deanp70 replied:
Removed all the updated imports, though I must say that when I try to install the package it tells me the specifically allowed versions don't work (using Python 3.8.5, so it might be a python version compatibility issue). If you're aware of this, then that's cool, but thought it might be important to modify the `setup.py` to support newer python versions

vijay120 replied:
Got it. We currently support 3.6 and 3.7, we are working on getting the backwards incompatible 3.8 version of MindMeld in soon. Thanks for revert.

vijay120 commented:
Hi Dean, I was testing out the CLI functionality and I see this, where access has to be granted to all private repos. Is there a way to grant perms to the one private repo that I want DAGsHub to read?

![Screen Shot 2021-06-07 at 12 18 12 PM](https://user-images.githubusercontent.com/766285/121075822-807caa00-c78a-11eb-8283-5189fdc2542d.png)

deanp70 commented:
@vijay120, sure. You selected the import option which we made to automate the process by scanning which repos you have on GitHub and let you pick which ones to import. If you know which repo and prefer to do with a URL, you can use the following link: https://dagshub.com/repo/connect/any

You then provide your username and access token and we'll only read the one repo you want to migrate. I hope I understood you and this is what you're looking for.

vijay120 commented:
@deanp70 Hi dean, I tried that but here is the issue I am facing: https://drive.google.com/file/d/1oJhaIzuVON74-TI_Cm_cJ5wQAfcyr_QW/view?usp=sharing . I still have to authorize access to all my other private repos.

deanp70 commented:
Hey @vijay120, thanks for the video, now everything is clear. Thanks for the feedback, you have stumbled on a bug, and we'll fix it ASAP. in the meantime, you can migrate your project, using the following link: https://dagshub.com/repo/migrate

It should look familiar to connect, but in order to make it equivalent, make sure to check the box next to "this repository will be a mirror". It should provide the exact same result (I checked that this works with one of my projects)

Sorry for the confusion, and thanks for pointing me toward the bug :)

deanp70 commented:
Hey, @vijay120 gentle reminder on this :)

deanp70 commented:
Hey @vijay120, any additional feedback for the PR, or could we perhaps merge it?

vijay120 commented on source/userguide/dvc.rst:
```diff
@@ -29,6 +36,7 @@ Available Options in MindMeld
 | --help          | Show the available options and their descriptions.                     |
 +-----------------+------------------------------------------------------------------------+
 
+[1] The DAGsHub central repository requires you `sign up to DAGsHub <https://dagshub.com/user/sign_up>`_ (it's completely free), and create a project – you can `connect an existing GitHub project <https://dagshub.com/repo/connect>`_ or `create a project from scratch <https://dagshub.com/repo/create>`_.
```
1. (it's completely free) -> (it's free for personal projects)
2. Could you please also mention https://dagshub.com/repo/migrate to migrate existing projects?
3. Could you also mention `setup_dagshub ` only works once  dvc project has been initialized in the repo?

deanp70 replied:
Done!

vijay120 commented:
Hi Dean, thanks for fixing the migration issue! That works now. I am currently facing this issue:

1. I migrated an exiting git project to dagshub
2. I added the dagshub remote to my local dir and checked out the dagshub branch
3. I committed the dvc models
4. When I tried to push to remote dagshub, I get this error:

```
(mindmeld) vijayrk@VIJAYRK-M-439M blueprints % git pull dags develop
hint: Pulling without specifying how to reconcile divergent branches is
hint: discouraged. You can squelch this message by running one of the following
hint: commands sometime before your next pull:
hint: 
hint:   git config pull.rebase false  # merge (the default strategy)
hint:   git config pull.rebase true   # rebase
hint:   git config pull.ff only       # fast-forward only
hint: 
hint: You can replace "git config" with "git config --global" to set a default
hint: preference for all repositories. You can also pass --rebase, --no-rebase,
hint: or --ff-only on the command line to override the configured default per
hint: invocation.
From https://dagshub.com/vijay120/blueprints
 * branch            develop    -> FETCH_HEAD
Already up to date.
(mindmeld) vijayrk@VIJAYRK-M-439M blueprints % git push dags develop
To https://dagshub.com/vijay120/blueprints.git
 ! [rejected]        develop -> develop (non-fast-forward)
error: failed to push some refs to 'https://dagshub.com/vijay120/blueprints.git'
```

5. So I did a `git push -f`, and that suppressed the error, but I still dont see my commit in the remote dagshub project in the "commits" section. However, the project does show it was "updated" 2 mins ago. So something got updates, but I am not sure what. I want to see the experiments from the model building on the dashboard.

deanp70 commented:
Hey @vijay120 thanks for the detailed explanation. Is this repo publicly available so I can try to reproduce what happened?
Also, what happens if you try to push again now?

I'll try to reproduce this on my end, though not clear why there was a need for force push.

deanp70 commented:
I also fixed the issues you mentioned in the CR

vijay120 commented:
@deanp70  whats your dagshub ID? I will add you as a collab

deanp70 commented:
@vijay120, it's Dean (benefits of being one of the first users 🤗)

deanp70 commented:
Hey @vijay120, hope all is well. Could you please add me to the repo so I can debug the issue you've been having? I can't seem to reproduce it anywhere else.

vijay120 commented:
@deanp70 done

deanp70 commented:
Hey @vijay120, thanks for making me a collaborator. I tried reproducing the issue you experienced, but with a "clean" clone of the project, I couldn't manage to do that. For me, everything worked, including using the new commands in the PR to push the models (`.generated/`) to DagsHub. This leads me to believe that it might've been a git issue due to having a modified local project and pulling a different state into the working directory.

I made you a collaborator to my version of the project, so you can see the results in action (I used the `home_assistant` blueprint).

Regarding seeing experiments in the dashboard, to make sure, if you're talking about seeing parameters in metrics in the Experiments tab, that would require either using `DagsHub Logger`, or having parameters and metrics in standard formats (for more info, [read here](https://dagshub.com/docs/reference/git-tracking/)) or using MLflow. In the current project, I don't see any code for this.

deanp70 commented:
Hey @vijay120, hope all is well.
Was wondering if you have any comments on the PR? Could this be merged in?

vijay120 commented:
Thanks for reproducing it. Approving and merging this PR.

vijay120 commented:
@deanp70 Could you please fix the linting issues? Looks like some of them are not related to your code, so just rebasing master should fix them.

deanp70 commented:
@vijay120 I rebased as you suggested, and fixed all linting issues I know how to (though most are unrelated to the code I wrote). The only error I can't really fix without a huge modification is the too many return statements – it seems that the limit before my code was 11 and I added 3 additional ones so it's outside the limit, but this is how all the code for the CLI is written. Not sure if the limit could be extended or not, but I don't think it makes sense to rewrite the entire CLI.py because of this. Let me know if I'm missing something :)

vijay120 commented:
@deanp70 I am not able to make the corrections since I cannot checkout your code and then publish the update here. Could you please do the following:

```
mindmeld/components/question_answerer.py:611:101: E501 line too long (101 > 100 characters)
mindmeld/models/model.py:53:1: F401 '.labels.LabelEncoder' imported but unused
mindmeld/models/model.py:53:1: F401 '.labels.EntityLabelEncoder' imported but unused
************* Module mindmeld.cli
mindmeld/cli.py:200:0: R0911: Too many return statements (14/11) (too-many-return-statements)
```

For 1) just make it into two lines
2) and 3) just remove the imports
For 4) just do `#pylint: disable=too-many-arguments`

deanp70 commented:
@vijay120 I see, I wasn't sure if I should disable things. Fixed all the points you made, and hopefully, now linting should work fine.

deanp70 commented:
Hey @vijay120, a gentle reminder on this. I think all the issues are fixed tests should be passing now.

deanp70 commented:
Hey @vijay120, a gentle reminder on this PR. Can we merge it in?

deanp70 commented:
Yay, this is awesome!
